core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java

```diff
@@ -4476,6 +4476,40 @@ public String getDefaultScheme() {
     }
   }
 
+  @Test
+  public void testApplyDefaultServiceConfigIfInitialNameResolutionFails() throws Exception {
+    LoadBalancerRegistry.getDefaultRegistry().register(mockLoadBalancerProvider);
+    channelBuilder.maxTraceEvents(10);
+    timer.forwardNanos(1234);
+    try {
+      FakeNameResolverFactory nameResolverFactory = new FakeNameResolverFactory
+              .Builder(expectedUri).setServers(Collections.singletonList(new
+              EquivalentAddressGroup(socketAddress))).build();
+      channelBuilder.nameResolverFactory(nameResolverFactory);
+      Map<String, Object> defaultServiceConfig =
+              parseConfig("{\"methodConfig\":[{"
+                      + "\"name\":[{\"service\":\"SimpleService1\"}],"
+                      + "\"waitForReady\":true}]}");
+      channelBuilder.defaultServiceConfig(defaultServiceConfig);
+      createChannel();
+      FakeNameResolverFactory.FakeNameResolver resolver = nameResolverFactory.resolvers.get(0);
+      Status resolutionError = Status.UNAVAILABLE
+              .withDescription("Initial Name Resolution error, using default service config");
+      resolver.listener.onError(resolutionError);
+      verify(mockLoadBalancer).handleNameResolutionError(resolutionError);
+      int prevSize = getStats(channel).channelTrace.events.size();
```

Comment:
You should store the trace events size before calling onError and get the events size again after calling onError and assert that it has increased by 1.

Reply:
@kannanjgithub Fixed this point but not sure while calling the onError() events size needs to be increased by 1 or decreased by 1 as in our UT case it's expecting to decrease by 1 otherwise assertion is failing.
Also, If I store the trace events size before calling onError() it expecting to decrease events size by 2.

Please confirm.

Reply:
What is the value of getStats(channel).channelTrace.events.size() before calling onError and after calling onError?

Reply:
What are scenario1 and scenario2? In the [file](https://github.com/vinodhabib/grpc-java/blob/904b89362c87991f3c52d4d764988c10538dbbe7/core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java) in the previous commit you had only one onError call, but now you have made two calls.

Reply:
After your changes in ManagedChannelImpl.java, when onError is called there will be 2 trace events now, one [here](https://github.com/grpc/grpc-java/blob/b51a622dfb540af7e64060fb512585460ec064df/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L1841) and one [here](https://github.com/grpc/grpc-java/blob/b51a622dfb540af7e64060fb512585460ec064df/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L1827). So the trace events count will become +2 after calling onError. You should add assertion for the contents of both trace events appended by the call to onError.

Reply:
By referring the onResult() UT cases was trying to add 2 onError() call as they were trying to do the same. I think only one onError call is enough in our case. I will revert back to same as earlier and check the events count once again. let you know the latest update.

As trying out the above with 2 onError calls in that context i was mentioned about scenario 1 and scenario 2 with before and after onError call.

Reply:
@kannanjgithub Fixed the Junit as per your suggestion and clean up the same along with trace events count to +2. Request you have a look once.